File: go/middleware/delegation_boundary_test.go
package middleware

import (
	"context"
	"testing"

	"github.com/soypete/pedro-agentware/go/tools"
)

// The delegation envelope carried across the middleware/delegation boundary
// (docs/tenant-proxy-reference.md): invoking_subject is first-class; the
// extended fields ride in CallerContext.Metadata so the envelope survives every
// hop and reaches the decision point and executor unchanged.
var delegationEnvelope = map[string]string{
	"organization": "acme-corp",
	"workspace":    "ws-sales-pipeline",
	"agent_id":     "agent-gh-bot",
	"connector_id": "github-connector",
	"capability":   "github.issue.write",
	"action":       "execute",
	"resource":     "github:repo:acme-corp/sales-pipeline",
	"trace_id":     "trace-2f8a9c",
	"approval_id":  "apr-2026-0007",
}

const (
	envelopeHumanSubject = "U_HUMAN"
	envelopeAgentID      = "agent-gh-bot"
)

func copyEnvelope() map[string]string {
	m := make(map[string]string, len(delegationEnvelope))
	for k, v := range delegationEnvelope {
		m[k] = v
	}
	return m
}

// The full envelope survives every delegation hop, and the child running under
// its own service id still carries the human subject.
func TestDelegationEnvelopeSurvivesEveryHop(t *testing.T) {
	human := CallerContext{
		UserID:          envelopeHumanSubject,
		InvokingSubject: envelopeHumanSubject,
		Metadata:        copyEnvelope(),
	}
	child := human.Delegate("span-1")
	child.UserID = envelopeAgentID
	grandchild := child.Delegate("span-2")
	grandchild.UserID = envelopeAgentID

	ctxs := []CallerContext{human, child, grandchild}
	depths := []int{0, 1, 2}
	for i, ctx := range ctxs {
		if ctx.DelegationDepth != depths[i] {
			t.Errorf("hop %d: DelegationDepth = %d, want %d", i, ctx.DelegationDepth, depths[i])
		}
		if ctx.InvokingSubject != envelopeHumanSubject {
			t.Errorf("hop %d: InvokingSubject = %q, want %q", i, ctx.InvokingSubject, envelopeHumanSubject)
		}
		for k, want := range delegationEnvelope {
			if got := ctx.Metadata[k]; got != want {
				t.Errorf("hop %d: metadata[%q] = %q, want %q", i, k, got, want)
			}
		}
	}
}

// Delegate() never rewrites the human subject; the subagent's own id is a
// different field (UserID), not an override of the attribution.
func TestDelegateNeverRewritesTheHumanSubject(t *testing.T) {
	parent := CallerContext{InvokingSubject: envelopeHumanSubject, Metadata: copyEnvelope()}

	child := parent.Delegate("span-1")
	child.UserID = envelopeAgentID

	if child.InvokingSubject != envelopeHumanSubject {
		t.Errorf("child InvokingSubject = %q, want %q", child.InvokingSubject, envelopeHumanSubject)
	}
	if parent.InvokingSubject != envelopeHumanSubject {
		t.Errorf("parent InvokingSubject = %q, must be untouched", parent.InvokingSubject)
	}
	if child.UserID != envelopeAgentID {
		t.Errorf("child UserID = %q, want %q", child.UserID, envelopeAgentID)
	}
}

// A missing caller is fail-closed: untrusted, empty delegation fields, and an
// empty envelope.
func TestMissingContextIsFailClosedWithEmptyEnvelope(t *testing.T) {
	caller := getCallerContext(context.Background())

	if caller.Trusted {
		t.Error("missing caller must default to Trusted=false")
	}
	if caller.InvokingSubject != "" || caller.ParentSpan != "" || caller.DelegationDepth != 0 {
		t.Errorf("expected empty delegation fields, got %+v", caller)
	}
	for k := range delegationEnvelope {
		if v, ok := caller.Metadata[k]; ok && v != "" {
			t.Errorf("missing context must not carry envelope field %q=%q", k, v)
		}
	}
}

// The envelope reaches both the policy decision point and the executor
// unchanged: the middleware is a pass-through, not a re-envelope.
func TestEnvelopeReachesDecisionPointAndExecutorUnchanged(t *testing.T) {
	var policyMeta, execMeta map[string]string

	exec := &mockExecutor{execFn: func(ctx context.Context, name string, args map[string]any) (*tools.Result, error) {
		if c, ok := CallerFromContext(ctx); ok {
			execMeta = c.Metadata
		}
		return &tools.Result{Success: true}, nil
	}}

	caller := CallerContext{
		UserID:          envelopeAgentID,
		InvokingSubject: envelopeHumanSubject,
		Metadata:        copyEnvelope(),
	}

	mw := NewMiddleware(exec).
		WithPolicy(&policyCaptureEvaluator{onEvaluate: func(name string, args map[string]any, c CallerContext) {
			policyMeta = c.Metadata
		}}).
		WithAuditor(&mockAuditor{})

	ctx := WithCallerContext(context.Background(), caller)
	if _, err := mw.Execute(ctx, delegationEnvelope["capability"], map[string]any{}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	for name, meta := range map[string]map[string]string{"policy": policyMeta, "executor": execMeta} {
		if meta == nil {
			t.Fatalf("caller metadata did not reach the %s boundary", name)
		}
		for k, want := range delegationEnvelope {
			if got := meta[k]; got != want {
				t.Errorf("%s boundary saw metadata[%q] = %q, want %q", name, k, got, want)
			}
		}
	}
}

type requireHumanEvaluator struct{}

func (e *requireHumanEvaluator) Evaluate(toolName string, args map[string]any, caller CallerContext) Decision {
	if caller.InvokingSubject == "" {
		return Decision{Action: ActionDeny, Rule: "require-human", Reason: "no human subject"}
	}
	return Decision{Action: ActionAllow, Rule: "human-ok"}
}

// A forged context -- envelope but no human subject -- is denied and audited as
// a denial; the record does not fabricate an attribution.
func TestForgedContextWithoutHumanSubjectIsDenied(t *testing.T) {
	forged := CallerContext{
		UserID: envelopeAgentID,
		Metadata: map[string]string{
			"agent_id": envelopeAgentID,
			"trace_id": "forged-trace",
		},
	}
	auditor := &mockAuditor{}
	mw := NewMiddleware(&mockExecutor{}).
		WithPolicy(&requireHumanEvaluator{}).
		WithAuditor(auditor)

	ctx := WithCallerContext(context.Background(), forged)
	if _, err := mw.Execute(ctx, "github.create_issue", map[string]any{}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(auditor.records) != 1 {
		t.Fatalf("expected 1 audit record, got %d", len(auditor.records))
	}
	if auditor.records[0].Decision != string(ActionDeny) {
		t.Errorf("forged context must be denied, got %+v", auditor.records[0])
	}
	if auditor.records[0].InvokingSubject != "" {
		t.Errorf("denial must not fabricate a subject, got %q", auditor.records[0].InvokingSubject)
	}
}

// A missing caller is denied by a policy that requires a human subject, and the
// record stays empty rather than being promoted.
func TestMissingContextIsDeniedAndNotPromoted(t *testing.T) {
	auditor := &mockAuditor{}
	mw := NewMiddleware(&mockExecutor{}).
		WithPolicy(&requireHumanEvaluator{}).
		WithAuditor(auditor)

	if _, err := mw.Execute(context.Background(), "github.create_issue", map[string]any{}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(auditor.records) != 1 {
		t.Fatalf("expected 1 audit record, got %d", len(auditor.records))
	}
	if auditor.records[0].Decision != string(ActionDeny) {
		t.Errorf("missing context must be denied, got %+v", auditor.records[0])
	}
}

// The audit record is the linkage point: the human subject and chain survive
// into the record even when the subagent runs under its own service id.
func TestAuditRecordPreservesHumanLinkageThroughTheChain(t *testing.T) {
	auditor := &mockAuditor{}
	mw := NewMiddleware(&mockExecutor{}).WithAuditor(auditor)

	human := CallerContext{UserID: envelopeHumanSubject, InvokingSubject: envelopeHumanSubject, Metadata: copyEnvelope()}
	sub := human.Delegate("sub-1")
	sub.UserID = envelopeAgentID

	ctx := WithCallerContext(context.Background(), sub)
	if _, err := mw.Execute(ctx, delegationEnvelope["capability"], map[string]any{}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if len(auditor.records) != 1 {
		t.Fatalf("expected 1 audit record, got %d", len(auditor.records))
	}
	rec := auditor.records[0]
	if rec.InvokingSubject != envelopeHumanSubject {
		t.Errorf("record InvokingSubject = %q, want %q", rec.InvokingSubject, envelopeHumanSubject)
	}
	if rec.ParentSpan != "sub-1" {
		t.Errorf("record ParentSpan = %q, want sub-1", rec.ParentSpan)
	}
	if rec.DelegationDepth != 1 {
		t.Errorf("record DelegationDepth = %d, want 1", rec.DelegationDepth)
	}
}

File: python/tests/delegation_boundary_test.py
"""Delegation-boundary contract tests: the proxy/approval delegation envelope.

These pin the agentware side of the delegation protocol documented in
``docs/tenant-proxy-reference.md``:

> a proxy/approval protocol carrying delegation context: ``invoking_subject``,
> ``agent_id``, tenant/workspace, ``trace_id``, idempotency key, approval id.

The original human subject is first-class on :class:`CallerContext`
(``invoking_subject``); the extended envelope (organization, workspace,
``agent_id``, ``connector_id``, capability, action, resource, ``trace_id`` and
the *optional* ``approval_id``) rides in ``CallerContext.metadata`` so it
survives every delegation hop and reaches the decision point and executor
unchanged.

The tests prove:

- **preservation** — every envelope field survives every ``delegate()`` hop
  verbatim, and a child cannot mutate the parent's envelope;
- **attribution** — authorization and audit resolve to the original human
  subject, never to the subagent's ``agent_id`` or a forged identity;
- **fail closed** — a missing caller, a missing envelope, or a forged context
  is never promoted to an allow; ``approval_id`` may be absent, required
  envelope fields may not.

No Kei invocation envelope, connector contract, migration, deployment, or
credential is modified: these tests assert the boundary that exists.
"""

import json
import sys
from pathlib import Path

# Point at THIS worktree's package (which contains `kei`), independent of CWD.
sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "src"))

import pytest

from pedro_agentware.kei import KeiProxyEvaluator
from pedro_agentware.middleware import (
    Action,
    AuditedToolClient,
    AuditFilter,
    CallerContext,
    Decision,
    InMemoryAuditor,
    MiddlewareImpl,
)
from pedro_agentware.middleware.policy import Policy, SimplePolicyEvaluator

ENVELOPE_PATH = Path(__file__).resolve().parent / "fixtures" / "kei" / "delegation-envelope-v1.json"
ENVELOPE = json.loads(ENVELOPE_PATH.read_text())

# Fields that ride in CallerContext.metadata. invoking_subject is first-class.
METADATA_ENVELOPE_FIELDS = {
    "organization",
    "workspace",
    "agent_id",
    "connector_id",
    "capability",
    "action",
    "resource",
    "trace_id",
    "approval_id",
}
# approval_id is the only optional envelope field.
REQUIRED_ENVELOPE_FIELDS = METADATA_ENVELOPE_FIELDS - {"approval_id"}


def _deny_all() -> SimplePolicyEvaluator:
    return SimplePolicyEvaluator(Policy(rules=[], default_deny=True))


def _human_context(**overrides):
    """A caller carrying the full delegation envelope at the human entry point."""
    metadata = {k: v for k, v in ENVELOPE.items() if k != "invoking_subject"}
    base = dict(
        user_id=ENVELOPE["invoking_subject"],
        invoking_subject=ENVELOPE["invoking_subject"],
        session_id="C123",
        metadata=metadata,
    )
    base.update(overrides)
    return CallerContext(**base)


class TestDelegationEnvelopePreserved:
    """The full envelope survives every delegation hop, verbatim."""

    def test_envelope_survives_every_delegation_hop(self):
        human = _human_context()
        child = human.delegate("span-1")
        grandchild = child.delegate("span-2")

        assert (human.delegation_depth, child.delegation_depth, grandchild.delegation_depth) == (
            0,
            1,
            2,
        )
        for depth, ctx in ((0, human), (1, child), (2, grandchild)):
            for field in METADATA_ENVELOPE_FIELDS:
                assert ctx.metadata[field] == ENVELOPE[field], (depth, field)
            assert ctx.invoking_subject == ENVELOPE["invoking_subject"]

    def test_delegate_copies_envelope_so_children_cannot_mutate_the_parent(self):
        human = _human_context()
        child = human.delegate("span-1")

        child.metadata["trace_id"] = "forged-trace"
        child.metadata["approval_id"] = "forged-approval"

        assert human.metadata["trace_id"] == ENVELOPE["trace_id"]
        assert human.metadata["approval_id"] == ENVELOPE["approval_id"]

    def test_envelope_survives_a_metadata_override_on_delegate(self):
        human = _human_context()
        child = human.delegate("span-1", metadata={**human.metadata, "subtask_id": "t1"})

        for field in METADATA_ENVELOPE_FIELDS:
            assert child.metadata[field] == ENVELOPE[field]
        assert child.metadata["subtask_id"] == "t1"
        assert child.invoking_subject == ENVELOPE["invoking_subject"]


class TestOriginalHumanSubjectPreserved:
    """Authorization and audit resolve to the human, never the subagent."""

    def test_agent_runs_under_its_own_id_but_the_human_survives(self):
        human = _human_context()
        sub = human.delegate(span="sub-1")
        sub.user_id = ENVELOPE["agent_id"]  # the subagent runs under its service id

        assert sub.invoking_subject == ENVELOPE["invoking_subject"]
        assert sub.metadata["agent_id"] == ENVELOPE["agent_id"]
        assert sub.user_id == ENVELOPE["agent_id"]

    def test_delegate_refuses_to_overwrite_the_human_subject(self):
        human = _human_context()

        forged = human.delegate("span-1", invoking_subject="attacker@example.com")

        assert forged.invoking_subject == ENVELOPE["invoking_subject"]

    def test_metadata_cannot_smuggle_an_authoritative_subject(self):
        forged = _human_context()
        forged.metadata["invoking_subject"] = "attacker@example.com"

        # metadata is descriptive; the authoritative subject is first-class.
        assert forged.invoking_subject == ENVELOPE["invoking_subject"]


class TestMissingContextFailsClosed:
    """No caller: untrusted, empty envelope, no fabricated attribution."""

    def test_default_context_is_untrusted_with_an_empty_envelope(self):
        ctx = CallerContext()

        assert ctx.trusted is False
        assert ctx.invoking_subject == ""
        assert ctx.metadata == {}
        assert ctx.delegation_depth == 0

    @pytest.mark.asyncio
    async def test_client_without_a_caller_denied_by_a_deny_all_policy(self):
        client = AuditedToolClient(evaluator=_deny_all())

        with pytest.raises(PermissionError):
            await client.Execute("github.create_issue", {}, "U1", "C1", None, lambda **k: "never")

        assert client.records()[0].decision.action == Action.DENY

    @pytest.mark.asyncio
    async def test_client_without_a_caller_still_audits_as_the_entry_user(self):
        client = AuditedToolClient(source="test-harness")

        async def tool(**kwargs):
            return "ok"

        await client.Execute("echo", {"value": "hi"}, "U_HUMAN", "C123", None, tool)

        rec = client.records()[0]
        assert rec.invoking_subject == "U_HUMAN"  # human entry point
        assert rec.delegation_depth == 0
        assert rec.framework == "test-harness"


class TestForgedAndIncompleteContextFailsClosed:
    """A forged or incomplete envelope is never promoted to an allow."""

    def test_envelope_without_a_human_subject_is_not_authoritative(self):
        forged = CallerContext(
            user_id=ENVELOPE["agent_id"],
            metadata={
                "organization": ENVELOPE["organization"],
                "workspace": ENVELOPE["workspace"],
                "agent_id": ENVELOPE["agent_id"],
                "connector_id": ENVELOPE["connector_id"],
                "capability": ENVELOPE["capability"],
                "action": ENVELOPE["action"],
                "resource": ENVELOPE["resource"],
                "trace_id": "forged-trace",
                "approval_id": "forged-approval",
            },
        )

        assert forged.invoking_subject == ""
        assert forged.trusted is False

    @pytest.mark.asyncio
    async def test_forged_context_without_a_subject_is_denied_at_the_boundary(self):
        class RequireHuman:
            def evaluate(self, tool_name, args, caller):
                if not caller.invoking_subject:
                    return Decision(
                        action=Action.DENY, rule="require-human", reason="no human subject"
                    )
                return Decision(action=Action.ALLOW, rule="human-ok")

        forged = CallerContext(
            user_id=ENVELOPE["agent_id"],
            metadata={"agent_id": ENVELOPE["agent_id"], "trace_id": "forged-trace"},
        )

        client = AuditedToolClient(evaluator=RequireHuman())
        with pytest.raises(PermissionError, match="no human subject"):
            await client.Execute(
                "github.create_issue",
                {},
                forged.user_id,
                "C1",
                None,
                lambda **k: "never",
                caller=forged,
            )

        assert client.records()[0].decision.action == Action.DENY

    @pytest.mark.asyncio
    async def test_missing_trace_id_fails_closed_but_approval_id_is_optional(self):
        class RequireTrace:
            def evaluate(self, tool_name, args, caller):
                if not caller.metadata.get("trace_id"):
                    return Decision(action=Action.DENY, rule="require-trace", reason="no trace_id")
                return Decision(action=Action.ALLOW, rule="trace-ok")

        client = AuditedToolClient(evaluator=RequireTrace())

        missing_trace = _human_context()
        missing_trace.metadata.pop("trace_id")
        with pytest.raises(PermissionError, match="no trace_id"):
            await client.Execute(
                "t",
                {},
                missing_trace.user_id,
                "C1",
                None,
                lambda **k: "never",
                caller=missing_trace,
            )

        # approval_id is optional: the full envelope minus approval_id passes.
        ok = _human_context()
        ok.metadata.pop("approval_id")
        assert "approval_id" not in ok.metadata
        result = await client.Execute("t", {}, ok.user_id, "C1", None, lambda **k: "ran", caller=ok)
        assert result == "ran"

    @pytest.mark.asyncio
    async def test_forged_metadata_cannot_stand_in_for_a_missing_required_field(self):
        """A forged approval/trace key in metadata never fabricates the real one."""
        missing_trace = _human_context()
        missing_trace.metadata.pop("trace_id")
        missing_trace.metadata["approval_id"] = "forged-approval"

        assert missing_trace.metadata.get("trace_id") is None
        assert missing_trace.invoking_subject == ENVELOPE["invoking_subject"]


class TestProxyBoundaryCarriesEnvelopeAndAuthorizesTheHuman:
    """The envelope reaches the decision point; the proxy authorizes the human."""

    class RecordingProxyEvaluator(KeiProxyEvaluator):
        def __init__(self, client):
            super().__init__(client)
            self.seen_caller: CallerContext | None = None

        def evaluate(self, tool_name, args, caller):
            self.seen_caller = caller
            return super().evaluate(tool_name, args, caller)

    @pytest.mark.asyncio
    async def test_envelope_reaches_the_decision_point_unchanged(self):
        class FakeProxy:
            def __init__(self):
                self.calls: list[dict] = []

            def authorize(self, user_id, tool, action, resource):
                self.calls.append(
                    {"user_id": user_id, "tool": tool, "action": action, "resource": resource}
                )
                return {"decision": "permit"}

        proxy = FakeProxy()
        evaluator = self.RecordingProxyEvaluator(proxy)

        human = _human_context()
        sub = human.delegate("sub-1")
        sub.user_id = ENVELOPE["agent_id"]

        client = AuditedToolClient(evaluator=evaluator)
        await client.Execute(
            ENVELOPE["capability"],
            {
                "owner": "acme-corp",
                "repo": "sales-pipeline",
                "action": ENVELOPE["action"],
            },
            sub.user_id,
            "C1",
            None,
            lambda **k: "ok",
            caller=sub,
        )

        seen = evaluator.seen_caller
        assert seen is not None
        assert seen.invoking_subject == ENVELOPE["invoking_subject"]
        for field in METADATA_ENVELOPE_FIELDS:
            assert seen.metadata[field] == ENVELOPE[field], field

        # The proxy boundary authorizes the human, not the subagent's id.
        assert proxy.calls[0]["user_id"] == ENVELOPE["invoking_subject"]
        assert proxy.calls[0]["tool"] == ENVELOPE["capability"]
        assert proxy.calls[0]["action"] == ENVELOPE["action"]
        assert proxy.calls[0]["resource"] == ENVELOPE["resource"]

    @pytest.mark.asyncio
    async def test_non_affirmative_decision_denies_even_with_the_full_envelope(self):
        class _Client:
            def authorize(self, user_id, tool, action, resource):
                return {"decision": "deny"}

        client = AuditedToolClient(evaluator=KeiProxyEvaluator(_Client()))
        human = _human_context()

        with pytest.raises(PermissionError):
            await client.Execute(
                "github.read", {}, human.user_id, "C1", None, lambda **k: "never", caller=human
            )

        rec = client.records()[0]
        assert rec.decision.action == Action.DENY
        # A denial still audits as the human.
        assert rec.invoking_subject == ENVELOPE["invoking_subject"]

    @pytest.mark.asyncio
    async def test_unreachable_proxy_denies_even_with_the_full_envelope(self):
        class _Broken:
            def authorize(self, user_id, tool, action, resource):
                raise ConnectionError("proxy unreachable")

        client = AuditedToolClient(evaluator=KeiProxyEvaluator(_Broken()))
        human = _human_context()

        with pytest.raises(PermissionError, match="proxy unreachable"):
            await client.Execute(
                "github.read", {}, human.user_id, "C1", None, lambda **k: "never", caller=human
            )

        assert client.records()[0].decision.action == Action.DENY


class TestAuditRecordCarriesDelegationLinkage:
    """The audit record is the linkage point: human subject + chain + envelope."""

    @pytest.mark.asyncio
    async def test_audit_record_links_the_human_through_the_chain(self):
        client = AuditedToolClient(source="test-harness")
        human = _human_context()
        sub = human.delegate("sub-1")
        sub.user_id = ENVELOPE["agent_id"]

        async def tool(**kwargs):
            return "ok"

        await client.Execute("echo", {}, sub.user_id, "C1", None, tool, caller=sub)

        rec = client.records()[0]
        assert rec.invoking_subject == ENVELOPE["invoking_subject"]
        assert rec.parent_span == "sub-1"
        assert rec.delegation_depth == 1
        assert rec.framework == "test-harness"

    def test_middleware_impl_preserves_the_envelope_to_the_policy(self):
        captured: dict = {}

        class CapturingEvaluator:
            def evaluate(self, tool_name, args, caller):
                captured["caller"] = caller
                return Decision(action=Action.ALLOW, reason="capture")

        class EchoExecutor:
            def execute(self, tool_name, args):
                return ({"echo": args.get("value")}, True, "")

        auditor = InMemoryAuditor()
        mw = MiddlewareImpl(EchoExecutor(), auditor=auditor, evaluator=CapturingEvaluator())

        sub = _human_context().delegate("sub-1")
        sub.user_id = ENVELOPE["agent_id"]
        mw.execute("echo", {"value": "hi"}, sub)

        seen = captured["caller"]
        assert seen.invoking_subject == ENVELOPE["invoking_subject"]
        for field in METADATA_ENVELOPE_FIELDS:
            assert seen.metadata[field] == ENVELOPE[field], field

        record = auditor.query(AuditFilter())[0]
        assert record.invoking_subject == ENVELOPE["invoking_subject"]
        assert record.delegation_depth == 1
